### PR TITLE
WIP: tests passing in Mocha (Node), same tests in Cypress (browser)

### PR DESCRIPTION
**Pull request by bahmutov**
All tests use same syntax and run under Node (Mocha) and all in browser (Cypress). Only thing not working are loading fixtures.

Main changes:

- use `expect` instead of `should`
- use `cy.stub` instead of `simple-mock` which is built into Cypress and is just `sinon`
- loading fixtures is work in progress, have to see how they are done, probably can abstract that as well
- all Node Mocha stuff is in single file [test/mocha-options.js](test/mocha-options.js) which makes every spec file much simpler

```js
require('../mocha-options')
const GitHub = require('../../')
// tests
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -16,3 +16,4 @@ doc/
 
 # browser builds
 dist/
+cypress/videos
```
Comment (gr2m) [inline .gitignore]: as there is no UI, would it make sense to disable video (if possible)?

File: cypress/fixtures/example.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "Using fixtures to represent data",
+  "email": "hello@cypress.io",
+  "body": "Fixtures are a great way to mock data for responses to routes"
+}
```
Comment (gr2m) [inline cypress/fixtures/example.json]: is that file needed?

File: cypress/plugins/index.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+// https://on.cypress.io/plugins-guide
+// ***********************************************************
+
+// This function is called when a project is opened or re-opened (e.g. due to
+// the project's config changing)
+
+module.exports = (on, config) => {
+  // `on` is used to hook into various events Cypress emits
+  // `config` is the resolved Cypress config
+}
```
Comment (gr2m) [inline cypress/plugins/index.js]: is that file needed?

File: cypress/support/commands.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+// -- This is a child command --
+// Cypress.Commands.add("drag", { prevSubject: 'element'}, (subject, options) => { ... })
+//
+//
+// -- This is a dual command --
+// Cypress.Commands.add("dismiss", { prevSubject: 'optional'}, (subject, options) => { ... })
+//
+//
+// -- This is will overwrite an existing command --
+// Cypress.Commands.overwrite("visit", (originalFn, url, options) => { ... })
```
Comment (gr2m) [inline cypress/support/commands.js]: is that file needed?

File: cypress/support/index.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+//
+// You can read more here:
+// https://on.cypress.io/configuration
+// ***********************************************************
+
+// Import commands.js using ES2015 syntax:
+import './commands'
+
+// Alternatively you can use CommonJS syntax:
+// require('./commands')
```
Comment (gr2m) [inline cypress/support/index.js]: in case we don’t need `commands.js`, can we avoid this file, too?

File: test/mocha-options.js [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+      stub: function () {
+        return sandbox.stub.apply(sandbox, arguments)
+      }
+    }
+  })
+
+  afterEach(() => {
+    sandbox.restore()
+  })
+}
```
Comment (gr2m) [inline test/mocha-options.js:24]: clever, I like that setup

**Review by gr2m — COMMENTED** [review]
Thanks so much for your help Gleb, I really appreciate it!

> loading fixtures is work in progress, have to see how they are done, probably can abstract that as well

I’d like to start / stop the fixtures server, so there will be no need for mocking. It’s currently setup like this:

https://github.com/gr2m/octokit-rest-browser-experimental/blob/feca70ec0d0623b73b0160a6e3626dc4d0d987ba/package.json#L80-L82

But that’s not very elegant, I wonder if you have a better idea?

File: package.json [hunk 7]
```diff
@@ -50,6 +50,7 @@
     "browserify": "^15.2.0",
     "chai": "^4.1.2",
     "coveralls": "^3.0.0",
+    "cypress": "^1.4.1",
```
Comment (gr2m) [inline package.json:53]: where is cypress downloading its binary to on travis? I’d add that folder to caching

**Comment by bahmutov** [comment]
To node_modules

**Comment by gr2m** [comment]
> To node_modules

I don’t want to cache the entire node_modules folder, that caused too much headache in the past. Can you tell what folder exactly?

**Comment by bahmutov** [comment]
Node_modules/cypress but do you cache node_modules/.bin also ?

**Comment by bahmutov** [comment]
I removed cypress support files and disabled video recording.

**Comment by bahmutov** [comment]
for fixture server - what would it serve? I like to start my servers and run test using https://github.com/bahmutov/start-server-and-test which would avoid the `pretest` AND, OR shell script redirects. I think the fixture thing is the only thing that really needs to be done here :)

**Comment by bahmutov** [comment]
Ok, so for fixtures, since there are already in the `@octokit/fixtures` as files I hacked a little bit to get this to work
```js
// in spec files just load fixture ourselves
// const GitHubMock = fixtures.mock('api.github.com/markdown')
    const fix = require(`@octokit/fixtures/scenarios/api.github.com/markdown/normalized-fixture.json`)
    const GitHubMock = fixtures.mock(fix)
```
and in `@octokit/fixtures` module allow either passing fixtures or load them
```js
// @octokit/fixtures index.js
function get (name) {
  let fixtures
  if (typeof name === 'string') {
    fixtures = require(`./scenarios/${name}/normalized-fixture.json`)
  } else {
    fixtures = name
  }
  return fixtures.map(fixture => Object.assign({}, fixture))
}

function mock (fixturesOrName) {
  const fixtures = get(fixturesOrName)
  ....
```
what do you think? If we add to fixtures a way to use given fixtures, then we are golden and the tests will pass in both environments ...

File: .gitignore [hunk 8]
```diff
@@ -16,3 +16,9 @@ doc/
 
 # browser builds
 dist/
+
+# cypress test files
+cypress/fixtures
+cypress/plugins
+cypress/support
+cypress/videos
```
Comment (gr2m) [inline .gitignore:24]: could we just do `cypress/` instead?

Comment (bahmutov) [reply .gitignore:24]: yeah, we can!

**Comment by bahmutov** [comment]
Are you working on start-fixtures-server @gr2m ?

File: test/integration/deprecations-test.js [hunk 9]
```diff
@@ -20,29 +20,22 @@ describe('deprecations', () => {
       host: 'deprecations-test.com'
     })
     return github.integrations.getInstallations({})
-
     .then(() => {
-      console.warn.callCount.should.equal(2)
-
-      simple.restore()
+      expect(console.warn).to.be.calledTwice
```
Comment (gr2m) [inline test/integration/deprecations-test.js:24]: do you know of an alternative way to test for this that is not using a getter? This is currently failing the standard linting

Comment (gr2m) [reply test/integration/deprecations-test.js:24]: ah nevermind, this works:

```js
expect(console.warn.callCount).to.equal(2)
```

**Comment by bahmutov** [comment]
Yeah it works but would not give you a nicer error ;)

File: cypress.json [hunk 10]
```diff
@@ -0,0 +1,4 @@
+{
+  "integrationFolder": "test/integration",
+  "videoRecording": false
+}
```
Comment (gr2m) [inline cypress.json:4]: I try to avoid more configuration files if possible so I tried this

```
cypress run --config integrationFolder='test/scenarios',videoRecording=false
```

but I still get the error 

```
We looked but did not find a cypress.json file in this folder: /Users/gregor/Projects/octokit/rest
```

Is there a way to work around that?

**Comment by gr2m** [comment]
> Are you working on start-fixtures-server @gr2m ?

In the master branch, it is starting it in a background process:
https://github.com/gr2m/octokit-rest-browser-experimental/blob/master/.travis.yml#L29

And in the instructions for local testing I’d just mention that the server needs to be started, like this:
https://github.com/gr2m/octokit-rest-browser-experimental/tree/be4f8937051c897ff65bf451826434ddca0e1765#testing-with-puppeteer

**Comment by bahmutov** [comment]
I'm afraid it needs cypress.json file
